### PR TITLE
[Feature] Add YOLOXHead, SimOTAAssigner, and loss.

### PR DESCRIPTION
**Pull request by RangiLyu**
Add YOLOXHead, SimOTAAssigner, and loss.

### CODE REVIEW

File: mmdet/models/losses/iou_loss.py [hunk 1]
```diff
@@ -23,16 +23,25 @@ def iou_loss(pred, target, linear=False, eps=1e-6):
         target (torch.Tensor): Corresponding gt bboxes, shape (n, 4).
         linear (bool, optional): If True, use linear scale of loss instead of
             log scale. Default: False.
+        mode (str): Loss scaling mode, including linear, square, and log.
```
Comment (AronLin) [inline mmdet/models/losses/iou_loss.py:26]: 'linear' or \``'linear'``

File: mmdet/models/losses/iou_loss.py [hunk 2]
```diff
@@ -226,19 +235,26 @@ class IoULoss(nn.Module):
...
 
     Args:
-        linear (bool): If True, use linear scale of loss instead of log scale.
-            Default: False.
+        linear (bool): If True, use linear scale of loss else determined
+            by mode. Default: False.
         eps (float): Eps to avoid log(0).
         reduction (str): Options are "none", "mean" and "sum".
         loss_weight (float): Weight of loss.
+        mode (str): Loss scaling mode, including linear, square, and log.
```
Comment (AronLin) [inline mmdet/models/losses/iou_loss.py:243]: 'linear' or \``'linear'``

File: mmdet/models/losses/iou_loss.py [hunk 3]
```diff
@@ -23,16 +23,25 @@ def iou_loss(pred, target, linear=False, eps=1e-6):
...
+            Default: 'log'
         eps (float): Eps to avoid log(0).
 
     Return:
         torch.Tensor: Loss tensor.
     """
-    ious = bbox_overlaps(pred, target, is_aligned=True).clamp(min=eps)
+    assert mode in ['linear', 'square', 'log']
     if linear:
```
Comment (AronLin) [inline mmdet/models/losses/iou_loss.py:34]: Do we need to add deprecated warning?

File: mmdet/core/bbox/assigners/sim_ota_assigner.py [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+        # both in boxes and centers, shape: [num_fg, num_gt]
+        is_in_boxes_and_centers = (
+            is_in_gts[is_in_gts_or_centers, :]
+            & is_in_cts[is_in_gts_or_centers, :])
+        return is_in_gts_or_centers, is_in_boxes_and_centers
+
+    def dynamic_k_matching(self, cost, pair_wise_ious, num_gt, valid_mask):
+        matching_matrix = torch.zeros_like(cost)
+        # select candidate topk ious for dynamic-k calculation
+        topk_ious, _ = torch.topk(pair_wise_ious, self.candidate_topk, dim=0)
```
Comment (ZwwWayne) [inline mmdet/core/bbox/assigners/sim_ota_assigner.py:229]: pair_wise ->pairwise

File: mmdet/models/dense_heads/yolox_head.py [hunk 5]
```diff
@@ -0,0 +1,477 @@
...
+
+        assign_result = self.assigner.assign(
+            cls_preds.sigmoid() * objectness.unsqueeze(1).sigmoid(),
+            offset_priors, decoded_bboxes, gt_bboxes, gt_labels)
+
+        sampling_result = self.sampler.sample(assign_result, priors, gt_bboxes)
+        pos_inds = sampling_result.pos_inds
+        num_pos_per_img = pos_inds.size(0)
+
+        pred_ious_this_matching = assign_result.max_overlaps[pos_inds]
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/yolox_head.py:453]: this name can be shorter, do we need to highlight this matching?
